src/main/java/store/model/service/StoreService.java

```diff
@@ -0,0 +1,165 @@
+        boolean isActivePromotion = isPromotionActive(promotion);
+
+        int promotionStock = productRepository.findStock(productName, ProductType.PROMOTION);
+        return createShoppingCartCheck(productName, productPrice, requestedQuantity, promotion, isActivePromotion, promotionStock);
+    }
+
+    private Product getProductInfoOrThrow(String productName) {
+        Product product = getProductInfo(productName);
+        if (product == null) {
+            throw new IllegalStateException(ErrorMessage.NON_EXISTENT_PRODUCT.getMessage());
+        }
+        return product;
+    }
+
+    private ShoppingCartCheck createShoppingCartCheck(String productName, int productPrice, int requestedQuantity, Promotion promotion, boolean isActivePromotion, int promotionStock) {
+        if (isPromotionUnavailable(promotion, promotionStock)) {
+            return createNonPromotionShoppingCartCheck(productName, productPrice, requestedQuantity, isActivePromotion);
+        }
+        return createPromotionShoppingCartCheck(productName, productPrice, requestedQuantity, promotionStock, promotion, isActivePromotion);
+    }
+
+    private ShoppingCartCheck createNonPromotionShoppingCartCheck(String productName, int productPrice, int requestedQuantity, boolean isActivePromotion) {
+        return new ShoppingCartCheck(
+                productName,
+                productPrice,
+                requestedQuantity,
+                requestedQuantity,
+                0,
+                requestedQuantity,
+                0,
+                false,
+                isActivePromotion
+        );
+    }
+
+    private ShoppingCartCheck createPromotionShoppingCartCheck(String productName, int productPrice, int requestedQuantity, int promotionStock, Promotion promotion, boolean isActivePromotion) {
+        int buy = promotion.getBuy();
+        int get = promotion.getGet();
+
+        int regularCount = Math.max(requestedQuantity - promotionStock, 0);
+        int promotionCount = requestedQuantity - regularCount;
+
+        int fullPriceCount = calculateFullPriceCount(promotionCount, buy, get, regularCount);
+        int freeCount = calculateFreeCount(promotionCount, buy, get);
+        boolean canReceiveAdditionalFree = calculateAdditionalFree(requestedQuantity, buy, get, regularCount, promotionStock);
+
+        return new ShoppingCartCheck(productName, productPrice, requestedQuantity, regularCount, promotionCount, fullPriceCount, freeCount, canReceiveAdditionalFree, isActivePromotion);
+    }
+
+    private Product getProductInfo(String productName) {
+        Product product = productRepository.findInfo(productName, ProductType.PROMOTION);
+        if (product == null) {
+            product = productRepository.findInfo(productName, ProductType.REGULAR);
+        }
+        return product;
+    }
+
+    private Promotion getPromotionInfo(Product product) {
+        if (product != null && product.getPromotion() != null) {
```

Comment:
product.getPromotion() != null 도 괜찮지만
product 자체에 hasPromotion이라는 boolean 반환 함수를 만드는 것은 어떤가요?
```suggestion
        if (product != null && product.hasPromotion()) {
```

Reply:
감사합니다. 참고하겠습니다!